### PR TITLE
Fix/39039 duplicate ids

### PR DESCRIPTION
**Pull request by lucatume**
Ticket: removed to avoid bot closing the PR.

This PR fixes the duplicate attachments when referencing images by URL.

I'd like to have some thought on the lazy fetching I'm doing in the `Featured_Image_Uploader::get_attachment_ID_from_url` method; my thought is that a caching plugin in place should not bother the user.

But it's the eternal struggle between one big query and many small one.

Please note that featured images are imported for venues and organizers even when PRO is not active, just not shown.

### CODE REVIEW

**Comment by tr1b0t** [comment]
Closing this Pull Request because it is pull requested against the wrong branch. The Central ticket is for *Current Release*, but this was pull requested against *release/4.1.1*.

**Comment by lucatume** [comment]
Ticket: https://central.tri.be/issues/39039

File: src/Tribe/Importer/Featured_Image_Uploader.php [hunk 1]
```diff
@@ -86,8 +97,50 @@ private function upload_file( $file_url ) {
...
+		if ( $this->_attachment_guids === false ) {
+			/** @var \wpdb $wpdb */
+			global $wpdb;
+			$guids = $wpdb->get_results( "SELECT ID, guid FROM $wpdb->posts where post_type = 'attachment'" );
+
+			$this->_attachment_guids = $guids ? array_combine( wp_list_pluck( $guids, 'guid' ), wp_list_pluck( $guids, 'ID' ) ) : array();
+		}
+	}
+
+	protected function maybe_fetch_all_attacment_original_urls() {
```
Comment (borkweb) [inline src/Tribe/Importer/Featured_Image_Uploader.php]: There's a typo in this function name. `attacment` vs `attachment`

**Comment by barryhughes** [comment]
For ref, [team chat here](https://tribe.slack.com/archives/products-dev/p1458922849001075) covers why we're targeting 4.1.1 (though the ticket is for CR/4.2).

File: src/Tribe/Importer/Featured_Image_Uploader.php [hunk 2]
```diff
@@ -14,6 +14,16 @@ class Tribe__Events__Importer__Featured_Image_Uploader {
 	protected $featured_image;
 
 	/**
+	 * @var bool|array
+	 */
+	protected static $_attachment_guids_cache = false;
```
Comment (barryhughes) [inline src/Tribe/Importer/Featured_Image_Uploader.php]: Not a big problem, but I think we've generally been avoiding the `_prefix` convention, the access modifiers are enough by themselves.

File: src/Tribe/Importer/Featured_Image_Uploader.php [hunk 3]
```diff
@@ -86,8 +105,54 @@ private function upload_file( $file_url ) {
...
 
 		return $id;
 	}
 
+	protected function get_attachment_ID_from_url( $featured_image ) {
+		$this->maybe_init_attachment_guids_cache();
+		$this->maybe_init_attachment_original_urls_cache();
+
+		/** @var \wpdb $wpdb */
+		global $wpdb;
```
Comment (barryhughes) [inline src/Tribe/Importer/Featured_Image_Uploader.php]: Doesn't seem to be used - remove?

**Comment by barryhughes** [comment]
> But it's the eternal struggle between one big query and many small one.

Yeah I'm not sure what the best way forward is here, but I think we could run with this and see what happens once it's out in the wild.

**Comment by barryhughes** [comment]
[Merging in Luca's absence.](https://tribe.slack.com/archives/products-dev/p1459279516001269)

**Pull request merged**